File: services/api/internal/agent/bedrock.go
package agent

import (
	"context"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/bedrockruntime"
	"github.com/aws/aws-sdk-go-v2/service/bedrockruntime/document"
	"github.com/aws/aws-sdk-go-v2/service/bedrockruntime/types"
)

// defaultModelID is the granted entitlement on this account (Claude Haiku 4.5), also the cheapest,
// which matters under the low daily-token quota. Override with AGENTS_MODEL_ID. The cross-region
// inference-profile id (us.anthropic. prefix) is required; the bare anthropic.claude-* id raises
// ValidationException (on-demand throughput is not supported for these models).
const defaultModelID = "us.anthropic.claude-haiku-4-5-20251001-v1:0"

// BedrockConverse is the real Converser over the Bedrock Converse API.
type BedrockConverse struct {
	client    *bedrockruntime.Client
	modelID   string
	maxTokens int32
}

// NewBedrockConverse builds the client from the default AWS credential chain (the Fargate task role
// assumes the Bedrock-invoke principal). It does not make a network call, so a missing quota surfaces
// only when Audit runs.
func NewBedrockConverse(ctx context.Context) (*BedrockConverse, error) {
	cfg, err := awsconfig.LoadDefaultConfig(ctx)
	if err != nil {
		return nil, err
	}
	model := os.Getenv("AGENTS_MODEL_ID")
	if model == "" {
		model = defaultModelID
	}
	return &BedrockConverse{client: bedrockruntime.NewFromConfig(cfg), modelID: model, maxTokens: 512}, nil
}

// Converse runs one Converse turn, translating the agent's transcript types to and from the SDK.
func (b *BedrockConverse) Converse(
	ctx context.Context, system string, messages []Message, tools []ToolSpec,
) (Result, error) {
	in := &bedrockruntime.ConverseInput{
		ModelId:         aws.String(b.modelID),
		System:          []types.SystemContentBlock{&types.SystemContentBlockMemberText{Value: system}},
		Messages:        toSDKMessages(messages),
		InferenceConfig: &types.InferenceConfiguration{MaxTokens: aws.Int32(b.maxTokens), Temperature: aws.Float32(0)},
	}
	if len(tools) > 0 {
		in.ToolConfig = toSDKToolConfig(tools)
	}
	out, err := b.client.Converse(ctx, in)
	if err != nil {
		return Result{}, err
	}
	return parseSDKOutput(out), nil
}

func toSDKMessages(messages []Message) []types.Message {
	out := make([]types.Message, 0, len(messages))
	for _, m := range messages {
		role := types.ConversationRoleUser
		if m.Role == RoleAssistant {
			role = types.ConversationRoleAssistant
		}
		content := make([]types.ContentBlock, 0, len(m.Blocks))
		for _, blk := range m.Blocks {
			switch {
			case blk.ToolUse != nil:
				content = append(content, &types.ContentBlockMemberToolUse{Value: types.ToolUseBlock{
					ToolUseId: aws.String(blk.ToolUse.ID),
					Name:      aws.String(blk.ToolUse.Name),
					Input:     document.NewLazyDocument(blk.ToolUse.Input),
				}})
			case blk.ToolResult != nil:
				content = append(content, &types.ContentBlockMemberToolResult{Value: types.ToolResultBlock{
					ToolUseId: aws.String(blk.ToolResult.ToolUseID),
					Content: []types.ToolResultContentBlock{
						&types.ToolResultContentBlockMemberJson{Value: document.NewLazyDocument(blk.ToolResult.JSON)},
					},
				}})
			default:
				content = append(content, &types.ContentBlockMemberText{Value: blk.Text})
			}
		}
		out = append(out, types.Message{Role: role, Content: content})
	}
	return out
}

func toSDKToolConfig(tools []ToolSpec) *types.ToolConfiguration {
	specs := make([]types.Tool, 0, len(tools))
	for _, t := range tools {
		specs = append(specs, &types.ToolMemberToolSpec{Value: types.ToolSpecification{
			Name:        aws.String(t.Name),
			Description: aws.String(t.Description),
			InputSchema: &types.ToolInputSchemaMemberJson{Value: document.NewLazyDocument(t.InputSchema)},
		}})
	}
	return &types.ToolConfiguration{Tools: specs}
}

func parseSDKOutput(out *bedrockruntime.ConverseOutput) Result {
	res := Result{StopReason: string(out.StopReason)}
	msgOut, ok := out.Output.(*types.ConverseOutputMemberMessage)
	if !ok {
		return res
	}
	assistant := Message{Role: RoleAssistant}
	for _, block := range msgOut.Value.Content {
		switch v := block.(type) {
		case *types.ContentBlockMemberText:
			res.Text += v.Value
			assistant.Blocks = append(assistant.Blocks, Block{Text: v.Value})
		case *types.ContentBlockMemberToolUse:
			var input map[string]any
			if v.Value.Input != nil {
				_ = v.Value.Input.UnmarshalSmithyDocument(&input)
			}
			tu := ToolUse{ID: aws.ToString(v.Value.ToolUseId), Name: aws.ToString(v.Value.Name), Input: input}
			res.ToolUses = append(res.ToolUses, tu)
			assistant.Blocks = append(assistant.Blocks, Block{ToolUse: &tu})
		}
	}
	res.Assistant = assistant
	return res
}

File: services/agents/src/erasure_agents/bedrock.py
"""A thin wrapper over the Bedrock Converse API.

Converser is an interface so the agents can be tested against a scripted fake with no AWS calls.
The real implementation targets the cross-region inference-profile id (prefix `us.anthropic.`),
because the bare `anthropic.claude-*` id raises ValidationException (on-demand throughput is not
supported for these models).
"""

import dataclasses
import os
from typing import Any, Protocol

# The granted entitlement on this account is Claude Haiku 4.5; it is also the cheapest, which
# matters under the low daily-token quota. Override with AGENTS_MODEL_ID if needed.
DEFAULT_MODEL_ID = "us.anthropic.claude-haiku-4-5-20251001-v1:0"


@dataclasses.dataclass(frozen=True)
class ToolSpec:
    """A tool advertised to the model (Converse toolConfig)."""

    name: str
    description: str
    input_schema: dict[str, Any]


@dataclasses.dataclass(frozen=True)
class ToolUse:
    """A tool the model asked to call."""

    tool_use_id: str
    name: str
    input: dict[str, Any]


@dataclasses.dataclass(frozen=True)
class ConverseResult:
    """The parsed result of one Converse turn."""

    stop_reason: str
    text: str
    tool_uses: list[ToolUse]
    assistant_message: dict[str, Any]  # the raw content blocks, to append to the running transcript


class Converser(Protocol):
    def converse(
        self,
        system: str,
        messages: list[dict[str, Any]],
        tools: list[ToolSpec] | None = None,
    ) -> ConverseResult: ...


def parse_converse_response(resp: dict[str, Any]) -> ConverseResult:
    """Parse a raw boto3 Converse response into a ConverseResult."""
    message = resp["output"]["message"]
    text_parts: list[str] = []
    tool_uses: list[ToolUse] = []
    for block in message.get("content", []):
        if "text" in block:
            text_parts.append(block["text"])
        elif "toolUse" in block:
            tu = block["toolUse"]
            tool_uses.append(ToolUse(tu["toolUseId"], tu["name"], tu.get("input", {}) or {}))
    return ConverseResult(
        stop_reason=resp.get("stopReason", ""),
        text="".join(text_parts).strip(),
        tool_uses=tool_uses,
        assistant_message=message,
    )


class BedrockConverse:
    """The real Converse client. The boto3 client is created lazily so importing this module (and
    running the mocked tests) never requires AWS credentials."""

    def __init__(
        self,
        model_id: str | None = None,
        region: str | None = None,
        max_tokens: int = 512,
        client: Any | None = None,
    ) -> None:
        self._model_id = model_id or os.environ.get("AGENTS_MODEL_ID", DEFAULT_MODEL_ID)
        self._region = region or os.environ.get("AWS_REGION", "us-east-1")
        self._max_tokens = max_tokens
        self._client = client

    def _bedrock(self) -> Any:
        if self._client is None:
            import boto3

            self._client = boto3.client("bedrock-runtime", region_name=self._region)
        return self._client

    def converse(
        self,
        system: str,
        messages: list[dict[str, Any]],
        tools: list[ToolSpec] | None = None,
    ) -> ConverseResult:
        kwargs: dict[str, Any] = {
            "modelId": self._model_id,
            "messages": messages,
            "system": [{"text": system}],
            "inferenceConfig": {"maxTokens": self._max_tokens, "temperature": 0.0},
        }
        if tools:
            kwargs["toolConfig"] = {
                "tools": [
                    {
                        "toolSpec": {
                            "name": t.name,
                            "description": t.description,
                            "inputSchema": {"json": t.input_schema},
                        }
                    }
                    for t in tools
                ]
            }
        resp = self._bedrock().converse(**kwargs)
        return parse_converse_response(resp)
